File: Laboratorios/Laboratorio2/lab2a.py
from sys import stdin

listadoLet = ["A","B","C","D","E","F","G","H","I","J","K","L","M","N","O","P","Q","R","S","T","U","V","W","X","Y","Z"]
listadoNum = [10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36]

def mapearNumeroInicial(numero,baseOrigen):
    numeroReal = []
    for x in range(len(numero)):
        pos = -1
        for y in range(0,baseOrigen-9):
            if(numero[x] == listadoLet[y]):
                pos = y
        if(pos != -1):
            numeroReal.append(listadoNum[pos])
        else:
            numeroReal.append(int(numero[x]))
    return numeroReal

def mapeoFinal(numero, baseDestino):
    numeroFinal = ""
    for x in range(len(numero)):
        pos = -1
        for y in range(0,baseDestino-9):
            if(numero[x] == listadoNum[y]):
                pos = y
        if(pos != -1):
            numeroFinal += listadoLet[pos]
        else:
            numeroFinal += str(numero[x])

    print(numeroFinal)

def aBaseDecimal(numero,baseOrigen):
    potencia = len(numero)
    numeroDecimal = []
    for i in range(len(numero)):
        potencia -= 1
        numeroDecimal.append(numero[i] * (baseOrigen**potencia))
    return numeroDecimal

def deBaseDecimal(numero,baseDestino):
    dividendo = 0
    x = len(numero)
    for i in range(len(numero)):
        x -= 1
        dividendo += numero[i] * (10**x)
    numeroFinal = []
    cociente = -1
    while(cociente != 0):
        completo = dividendo/baseDestino
        cociente = int(completo)
        residuo = dividendo % baseDestino
        numeroFinal.append(residuo)
        dividendo = cociente
    numeroFinal.reverse()
    return numeroFinal

def main():
    casos = int(stdin.readline())
    for x in range(casos):
        conver = stdin.readline().strip().split()
        baseOrigen = int(conver[0])
        baseDestino = int(conver[2])
        numeroOriginal = conver[1]
        numeroFinal = []
        numero = mapearNumeroInicial(numeroOriginal,baseOrigen)
        flag = False
        for i in range(len(numero)):
            if(numero[i] >= baseOrigen):
                flag = True
        if(baseOrigen < 2 or baseOrigen > 36):
            print("Error base origen")
        elif(flag):
            print("Error numero")
        elif(baseDestino < 2 or baseDestino > 36):
            print("Error base destino")
        elif(baseOrigen == baseDestino):
            print(numeroOriginal)
        else:
            if(baseOrigen == 10):
                numeroFinal = deBaseDecimal(numero,baseDestino)
            elif(baseOrigen != 10 and baseDestino != 10):
                temp = aBaseDecimal(numero,baseOrigen)
                l = []
                a = 0
                for z in range(len(temp)):
                    a += temp[z]
                l.append(a)
                numeroFinal = deBaseDecimal(l,baseDestino)
            else:
                numeroFinal = aBaseDecimal(numero,baseOrigen)
            mapeoFinal(numeroFinal, baseDestino)


main()

File: Laboratorios/Laboratorio2/lab2a.c
#include <stdio.h>
#include <math.h>
#include <string.h>

char listado[36] = {'0','1','2','3','4','5','6','7','8','9','A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z'};

int numValido(char numVal[8], int baseOrigen){
	int flag = 1;
	int a = 0; int x = 0;
	int len = strlen(numVal);
	while(x < len){
		a = numVal[x];
		if(numVal[x] > 57){
			a = numVal[x] - 7;
		}
		a -= 48;
		if(a >= baseOrigen){
			flag = 0;
			break; 
		}
		x++;
	}
	return flag;
}

long long aBaseDecimal(char numero[8], int baseOrigen){
    long long numeroDecimal = 0;
    int potencia = strlen(numero) - 1;
    
    char temp[8];
    int i = 0; int a = 0;
    while(i < 8){
    	a = numero[i];
    	if(a > 57){
    		a -= 7;
		}
		a -= 48;
		temp[i] = a;
    	i++;
	}
    
    i = 0;
    while(potencia >= 0){
        numeroDecimal += temp[i] * (pow(baseOrigen,potencia));
        i++; potencia--;
    }
    return numeroDecimal;
}

void deBaseDecimal(long long numero, int baseDestino){
	int n = 0;
	char casi[100] = {'\0'};
	int x = 0;
	if(numero == 0){
		casi[x] = 0;
	}
	while(numero != 0){
		n = numero % baseDestino;
		numero = (long long) numero/baseDestino;
		casi[x] = listado[n];
		x++;
	}
	
	char final[100] = {'\0'};
	int len = strlen(casi);
	x = 0;
	while(x < len){
		len --;
		final[x] = casi[len];
		x ++;
	}
	printf("%s \n", final);
}

void main(){
	int casos;
	scanf("%d", &casos);
	while(casos > 0){
		int baseOrigen;
		int baseDestino;
		char numeroOriginal[8];
		scanf("%d", &baseOrigen);
		scanf("%s", &numeroOriginal);
		scanf("%d", &baseDestino);
		if(baseOrigen < 2 || baseOrigen > 36){
			printf("Error base origen \n");
		}else{
			int numVal = numValido(numeroOriginal, baseOrigen);
			if(numVal == 0){
				printf("Error numero \n");
			}else{
				if(baseDestino < 2 || baseDestino > 36){
					printf("Error base destino \n");
				}else{
					if(baseOrigen == baseDestino){
						printf("%s \n", numeroOriginal);			
					}else{
						long long temp = aBaseDecimal(numeroOriginal,baseOrigen);
						if (baseDestino == 10){
							printf("%lld \n", temp);
						}else {
							if(temp == 0){
								printf("%lld \n",temp);
							}else{
								deBaseDecimal(temp,baseDestino);
							}
							
						}
					}
				}
			}
			
		}
		casos--;
	}
}
